core/templates/services/question-validation.service.ts

```diff
@@ -53,6 +53,34 @@ export class QuestionValidationService {
     );
   }
 
+  hasDuplicateRules(question: Question): boolean {
+    const answerGroups = question._stateData.interaction.answerGroups;
+    const ruleSet = new Set<string>();
+
+    for (const group of answerGroups) {
+      for (const rule of group.rules) {
+        const ruleHash = `${rule.type}:${this.normalizeRuleInputs(rule.inputs)}`;
+
+        if (ruleSet.has(ruleHash)) {
+          return true;
+        }
+        ruleSet.add(ruleHash);
+      }
+    }
+
+    return false;
+  }
+
+  normalizeRuleInputs(inputs: Record<string, unknown>): string {
```

Comment:
Forgot to mention -- if you have any members that are private, you can add "private" before them to indicate that they shouldn't be called directly from outside this service.